### ISSUE TITLE
[MIPS] Position Independent code isn't properly processed

### ISSUE THREAD

**Issue by nicolasnoble** [NationalSecurityAgency/ghidra#1736]
This is using ghidra version 9.1.2.

When compiling PIC code for MIPS, the compiler will emit `bgez zero, label` (which disassembles into the pseudo-instruction `b label`), and `bal function`. When in ghidra, the former isn't properly processed:

- `b label` is being disassembled as `bgez zero, label` instead of its pseudo-form.
- the analyzer still sees it as a CONDITIONAL_JUMP, instead of a BRANCH, which confuses the analysis flow, and produces a bunch of errors in the decompiler:

![image](https://user-images.githubusercontent.com/7281574/78949273-f2bb0700-7a7f-11ea-876d-e5591b71456a.png)

A good minimal reproduction case for this would be the attached file:
[mips-infinite-loop.gz](https://github.com/NationalSecurityAgency/ghidra/files/4459102/mips-infinite-loop.gz)

You can see the problems in this screenshot:

![image](https://user-images.githubusercontent.com/7281574/78949379-34e44880-7a80-11ea-9e2c-ba7893d05e23.png)

What's wrong in this picture:

- Instruction at 00000000 is considered "CONDITIONAL_JUMP" as instruction flow.
- The decompiler complains that it had to remove unreachable block.

Loading this test files is simply done using the MIPS:LE:32:default:default language:

![image](https://user-images.githubusercontent.com/7281574/78949723-4c700100-7a81-11ea-98b7-6c46c35cb89d.png)

**Comment by mumbel** [comment]
Did you change the `bgez zero,label` to `b label` in your binary to see if it changes how you think it should look?  ghidra knows the conditional always takes the CBRANCH, you can see in the PCODE that the condition is hard coded 1 (true) and not a variable.  

maybe missing your point with the infinite loop binary.  There is an unreachable block in your function as it can never execute the `jr; addi`, your function is 4 instructions, and the decompiler isn't processing dead code.  Look into tool options for `eliminate unreachable code`, maybe that's what you're looking for.

**Comment by nicolasnoble** [comment]
@mumbel `bgez zero, label` *is* `b label`. Some compilers will compile `b label` using `beq zero, zero, label`, some will compile using `bgez zero, label`. It's not the same generated bytes, but it's effectively the same pseudo-instruction.

The example I gave is very basic. The longer binary I have has a bunch of branches to serve as syscalls, and the disassembler just runaways across all of them as if it was a single codepath - which is wrong.

I'm not looing for `eliminate unreachable code`, ghidra doesn't currently properly recognize `bgez zero, label` as `b label`.

In fact, as far as I can see, the problem is here:

https://github.com/NationalSecurityAgency/ghidra/blob/f36d7a29da4220f6a559c2432984f5e5f36ec7b9/Ghidra/Processors/MIPS/data/languages/mips32Instructions.sinc#L50-L68

This is probably missing an entry along these lines:

```
# 0000 0100 0000 0001 iiii iiii iiii iiii 
:b Rel16                        is $(AMODE) & prime=4 & rs=0 & rt=0 & Rel16 {
    delayslot(1); 
    goto Rel16; 
}
```

**Comment by mumbel** [comment]
I am asking if you changed your binary in ghidra to be the instruction `b label` that is already implemented to see if always cbranch vs branch even matters. As you state it's not the same bytes and it's not the same actual instruction

(Also you need to change the bit pattern to match bgez with but with rs=0, all you changed was a comment)

**Comment by nicolasnoble** [comment]
I haven't changed my ghidra binary, no; not yet anyway. I'm currently in the process of trying to understand how that process works. Thank you for the quick hint about the fact this is only a comment, not the actual pattern that this is looking at.

**Comment by mumbel** [comment]
ctrl+g when your cursor is over the bgez, or there is a byte editor window if you know the exact bytes (ctrl+g does the math for you for labels though)

**Comment by nicolasnoble** [comment]
I managed to dump the relevant code fragments, and patch it. Here's an archive with both files: 
[code-fragments.zip](https://github.com/NationalSecurityAgency/ghidra/files/4462460/code-fragments.zip)

Loading the original file `code-fragment-bgez.bin` in ghidra using MIPS:LE:32:default:default and then immediately hitting the key 'D' will yield the following:

![image](https://user-images.githubusercontent.com/7281574/79007240-9bfd0e00-7b0f-11ea-84b2-1c24365e32d4.png)

Loading the patched file `code-fragment-beq.bin` the same way will yield the following:

![image](https://user-images.githubusercontent.com/7281574/79007302-bafba000-7b0f-11ea-96ed-cdbeeab5700f.png)

**Comment by mumbel** [comment]
not disagreeing that the CBRANCH vs BRANCH looks off there, but unfinished analysis could be part of the blame in your screenshots (disassembling everything/creating functions/creating thunks/etc...)

Are you aware of any documentation/specs for your compiler/chip that specifically call out the `bgez`/`b idiom`, `b` is described as an assembly idiom of `beq` in the mips32 manual?  Just wondering as there are a lot of instructions that could be interpreted as variants of the same pseudo instructions, it doesn't make sense to implement undocumented ones.

Also it appears there is maybe a COP (guessing) instruction that is not implemented, not finding it in the manual either, unimplemented instructions are going to cause decompiler issues for sure.

**Comment by nicolasnoble** [comment]
There are definitely lots of pseudo-instructions missing from the definitions in ghidra. I've seen lots and lots of variants while disassembling MIPS stuff. I don't necessarily have documentation, no. I don't know if it's because there are different compilers, or if some programmers did their own pseudo-instruction aliases. Most of the time however, it's not necessarily a big deal, as it's not causing troubles with the instruction flow. The branch one can be very annoying however, as there are cases where the analysis gets thrown off at the end of code and into data.

**Comment by mumbel** [comment]
yeah, totally agree with you there and based on your examples for the CBRANCH.

huh, I must be going crazy (maybe it was the first binary you linked) I could have sworn the CBRANCH had a 1 in the pcode op, but these are variables.  You might edit your issue title as something along the lines of `CBRANCH that are always true should be treated like BRANCH`

**Comment by nicolasnoble** [comment]
Sounds good. I was investigating at the same time, trying to understand why this PIC code was showing broken, and I had no idea at the time that the branch was the issue.

**Comment by tellowkrinkle** [comment]
Ran into this issue as well with a processor definition that was missing the `beq zero, zero` → `b` entry: https://github.com/beardypig/ghidra-emotionengine/pull/47

Always-true conditional branches also seem to make the decompiler reluctant to move code past them, leading to unnecessarily complicated decompiler output
